### PR TITLE
Different modes and values for padding in separable filter

### PR DESCRIPTION
**Pull request by ckolbPTB**

### CODE REVIEW

File: src/mrpro/utils/filters.py [hunk 1]
```diff
@@ -54,31 +61,47 @@ def uniform_filter_3d(
...
-    return uniform_filter(data, (z, y, x), axis=(-3, -2, -1))
+    return uniform_filter(data, (z, y, x), axis=(-3, -2, -1), pad_mode=pad_mode, pad_value=pad_value)
 
 
-def _filter_separable(x: torch.Tensor, kernels: Sequence[torch.Tensor], axis: Sequence[int]) -> torch.Tensor:
+def filter_separable(
+    x: torch.Tensor,
+    kernels: Sequence[torch.Tensor],
+    axis: Sequence[int],
+    pad_mode: Literal['constant', 'reflect', 'replicate', 'circular'] = 'constant',
```
Comment (fzimmermann89) [inline src/mrpro/utils/filters.py]: allow 'none' ?

File: src/mrpro/utils/filters.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
 """Spatial and temporal filters."""
 
-# Copyright 2023 Physikalisch-Technische Bundesanstalt
+# Copyright 2024 Physikalisch-Technische Bundesanstalt
```
Comment (fzimmermann89) [inline src/mrpro/utils/filters.py]: afaik this should be kept as 2023, as the file was initially written 2023 and we want to claim it from 2023 onwards...

File: src/mrpro/utils/filters.py [hunk 3]
```diff
@@ -29,6 +30,8 @@
 def uniform_filter_3d(
```
Comment (fzimmermann89) [inline src/mrpro/utils/filters.py]: can we remove this function completly?

File: src/mrpro/utils/filters.py [hunk 4]
```diff
@@ -54,31 +61,47 @@ def uniform_filter_3d(
...
     if len(axis) != len(kernels):
-        raise ValueError('Must provide matching length kernels and axis arguments. ')
+        raise ValueError('Must provide matching length kernels and axis arguments.')
 
     # normalize axis to allow negative indexing in input
     axis = tuple([a % x.ndim for a in axis])
     if len(axis) != len(set(axis)):
         raise ValueError(f'Axis must be unique. Normalized axis are {axis}')
 
+    # for pad_mode = constant and pad_value = 0, padding is done inside convolution, otherwise pad() is used.
```
Comment (fzimmermann89) [inline src/mrpro/utils/filters.py]: conside rmoving this comment inside the if statement and just writing 
```
if ...
# padding is done inside convolution
else:
# padding is done with pad() before the convolution
```

Comment (fzimmermann89) [reply src/mrpro/utils/filters.py]: this avoids having the same if else structure twice

File: src/mrpro/utils/filters.py [hunk 5]
```diff
@@ -92,9 +115,16 @@ def _filter_separable(x: torch.Tensor, kernels: Sequence[torch.Tensor], axis: Se
         # swapping the last axis and the axis to filter over
         idx[ax], idx[-1] = idx[-1], idx[ax]
         x = x.permute(idx)
-        x = torch.nn.functional.conv1d(
-            x.flatten(end_dim=-2)[:, None, :], kernel[None, None, :], padding='same'
-        ).reshape(x.shape)
+        x_shape = list(x.shape)
+        if pad_mode == 'none':
```
Comment (fzimmermann89) [inline src/mrpro/utils/filters.py]: remove

File: src/mrpro/utils/filters.py [hunk 6]
```diff
@@ -92,9 +115,16 @@ def _filter_separable(x: torch.Tensor, kernels: Sequence[torch.Tensor], axis: Se
...
+            x_shape[-1] -= len(kernel) - 1
+        # flatten first to allow for circular, replicate and reflection padding for arbitrary tensor size
+        x = x.flatten(end_dim=-2)
+        if padding_conv == 'valid' and pad_mode != 'none':
+            left_pad = (len(kernel) - 1) // 2
+            right_pad = (len(kernel) - 1) - left_pad
+            x = torch.nn.functional.pad(x, pad=(left_pad, right_pad), mode=pad_mode, value=pad_value)
+        x = torch.nn.functional.conv1d(x[:, None, :], kernel[None, None, :], padding=padding_conv).reshape(x_shape)
```
Comment (fzimmermann89) [inline src/mrpro/utils/filters.py]: x= torch.nn.functional.conv1d(x_flat[:, None, :], kernel[None, None, :], padding=padding_conv).reshape(x.shape[:-1,1)

File: src/mrpro/utils/filters.py [hunk 7]
```diff
@@ -92,9 +115,16 @@ def _filter_separable(x: torch.Tensor, kernels: Sequence[torch.Tensor], axis: Se
...
         idx[ax], idx[-1] = idx[-1], idx[ax]
         x = x.permute(idx)
-        x = torch.nn.functional.conv1d(
-            x.flatten(end_dim=-2)[:, None, :], kernel[None, None, :], padding='same'
-        ).reshape(x.shape)
+        x_shape = list(x.shape)
+        if pad_mode == 'none':
+            x_shape[-1] -= len(kernel) - 1
+        # flatten first to allow for circular, replicate and reflection padding for arbitrary tensor size
+        x = x.flatten(end_dim=-2)
```
Comment (fzimmermann89) [inline src/mrpro/utils/filters.py]: x_flat

File: src/mrpro/utils/filters.py [hunk 8]
```diff
@@ -92,9 +115,16 @@ def _filter_separable(x: torch.Tensor, kernels: Sequence[torch.Tensor], axis: Se
...
+        if padding_conv == 'valid' and pad_mode != 'none':
+            left_pad = (len(kernel) - 1) // 2
+            right_pad = (len(kernel) - 1) - left_pad
+            x = torch.nn.functional.pad(x, pad=(left_pad, right_pad), mode=pad_mode, value=pad_value)
```
Comment (fzimmermann89) [inline src/mrpro/utils/filters.py]: x_flat

**Review by fzimmermann89 — COMMENTED** [review]
can we remove the 3d filter?
this was kind of an interim solution, as we first only had the 3d uniform filter.
we can move the only to just use the general uniform filter I think

**Review by fzimmermann89 — APPROVED** [review]
